matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/content/VideoCompressor.kt

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2021 The Matrix.org Foundation C.I.C.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.matrix.android.sdk.internal.session.content
+
+import android.content.Context
+import com.abedelazizshe.lightcompressorlibrary.CompressionListener
+import com.abedelazizshe.lightcompressorlibrary.VideoCompressor
+import com.abedelazizshe.lightcompressorlibrary.VideoQuality
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.withContext
+import timber.log.Timber
+import java.io.File
+import java.util.UUID
+import javax.inject.Inject
+
+internal class VideoCompressor @Inject constructor(private val context: Context) {
+    suspend fun compress(videoFile: File,
+                         quality: VideoQuality = VideoQuality.MEDIUM,
+                         isMinBitRateEnabled: Boolean = false,
+                         keepOriginalResolution: Boolean = true): File {
+        return withContext(Dispatchers.IO) {
+            val job = Job()
+            val destinationFile = createDestinationFile()
+
+            // Sadly it does not return the Job, the API is not ideal
+            VideoCompressor.start(
+                    context = null,
+                    srcUri = null,
+                    srcPath = videoFile.path,
+                    destPath = destinationFile.path,
+                    listener = object : CompressionListener {
+                        override fun onProgress(percent: Float) {
+                            Timber.d("Compressing: $percent%")
+                        }
+
+                        override fun onStart() {
+                            Timber.d("Compressing: start")
+                        }
+
+                        override fun onSuccess() {
+                            Timber.d("Compressing: success")
+                            job.complete()
+                        }
+
+                        override fun onFailure(failureMessage: String) {
+                            Timber.d("Compressing: failure: $failureMessage")
+                            job.completeExceptionally(Exception(failureMessage))
+                        }
+
+                        override fun onCancelled() {
+                            Timber.d("Compressing: cancel")
+                            job.cancel()
```

Comment:
@Dominaezzz there is probably room for improvement here, on https://github.com/AbedElazizShe/LightCompressor API, but also on my workaround :)

Reply:
I'm slightly confused by the use of both `Job` and `Dispatchers.IO`, I think one of them has to go.
In reality this API should've exposed a flow but oh well, it's not too hard to wrap.

Reply:
I've added the support to track the progress, but I'm a bit confused too about how I can improve that. The Dispatchers.IO is here because we are creating a file, and the Job is here to make the call "synchronous". What do you suggest?

Reply:
Looking at https://natario1.github.io/Transcoder/docs/events , (unfortunately) transcoding is already done on a background thread, so you don't need to use `Dispatchers.IO`. You could also just limit it to just the `createDestinationFile()`.

(I can't wait until `ProgressListener` is replaced with `Flow`)

Reply:
Thanks for the feedback, I will change that. By the way the link you provide also shows me that there is a `successCode` in the `onTranscodeCompleted` fun. I did not noticed that. When the value is `SUCCESS_NOT_NEEDED` it appears that the target file is not written, so it explain a bug I observed today with a 0-lenght file. I will fix that too.